sqlite3.go

```diff
@@ -874,61 +901,62 @@ func (c *SQLiteConn) exec(ctx context.Context, query string, args []namedValue)
 			for i := range stmtArgs {
 				stmtArgs[i].Ordinal = i + 1
 			}
-			res, err = s.(*SQLiteStmt).exec(ctx, stmtArgs)
+			var err error
+			res, err = s.exec(ctx, stmtArgs)
 			if err != nil && err != driver.ErrSkip {
-				s.Close()
+				s.finalize()
 				return nil, err
 			}
 			start += na
 		}
-		tail := s.(*SQLiteStmt).t
-		s.Close()
-		if tail == "" {
+		s.finalize()
+		if tail == nil || *tail == '\000' {
 			if res == nil {
 				// https://github.com/mattn/go-sqlite3/issues/963
 				res = &SQLiteResult{0, 0}
 			}
 			return res, nil
 		}
-		query = tail
+		pquery = tail
 	}
 }
 
-type namedValue struct {
-	Name    string
-	Ordinal int
-	Value   driver.Value
-}
-
 // Query implements Queryer.
 func (c *SQLiteConn) Query(query string, args []driver.Value) (driver.Rows, error) {
-	list := make([]namedValue, len(args))
+	list := make([]driver.NamedValue, len(args))
 	for i, v := range args {
-		list[i] = namedValue{
+		list[i] = driver.NamedValue{
 			Ordinal: i + 1,
 			Value:   v,
 		}
 	}
 	return c.query(context.Background(), query, list)
 }
 
-func (c *SQLiteConn) query(ctx context.Context, query string, args []namedValue) (driver.Rows, error) {
+func (c *SQLiteConn) query(ctx context.Context, query string, args []driver.NamedValue) (driver.Rows, error) {
+	pquery := C.CString(query)
+	op := pquery // original pointer
+	defer C.free(unsafe.Pointer(op))
+
+	var stmtArgs []driver.NamedValue
+	var tail *C.char
+	s := new(SQLiteStmt) // escapes to the heap so reuse it
 	start := 0
 	for {
-		stmtArgs := make([]namedValue, 0, len(args))
-		s, err := c.prepare(ctx, query)
-		if err != nil {
-			return nil, err
+		*s = SQLiteStmt{c: c, cls: true} // reset
```

Comment:
You never set the finalizer, which means we could end up leaking memory.
I think it might be best not to attempt to be clever here and revert these changes. `SQLiteStmt` itself is pretty small (a few bytes), and overwriting something with a mutex fields is potentially dangerous.

Reply:
> You never set the finalizer, which means we could end up leaking memory.

`runtime.SetFinalizer` was never needed here. The only way for this to leak a prepared statement would be for this code to panic thus preventing `s.finalize()` from being called. But in that case the code is unreasonable and the program should exit. Additionally, there is some overhead to `runtime.SetFinalizer`.

> I think it might be best not to attempt to be clever here and revert these changes. SQLiteStmt itself is pretty small (a few bytes), and overwriting something with a mutex fields is potentially dangerous.

Fair and the number of allocations only really add up when the query contains multiple SQL statements (see below). The overwrite here is safe because we never copy the `SQLiteStmt` struct by value (which `go vet` via [golang/tools/go/analysis/passes/copylock ](https://github.com/golang/tools/tree/master/go/analysis/passes/copylock)).

Happy to change if you feel strongly about this.

Comparison between reusing the `SQLiteStmt` and initializing one for each iteration.
```
benchmark                               old ns/op     new ns/op     delta
BenchmarkSuite/BenchmarkExecStep-10     1146170       1171202       +2.18%

benchmark                               old allocs     new allocs     delta
BenchmarkSuite/BenchmarkExecStep-10     8004           9003           +12.48%

benchmark                               old bytes     new bytes     delta
BenchmarkSuite/BenchmarkExecStep-10     72104         120059        +66.51%
```

Reply:
> But in that case the code is unreasonable and the program should exit. 

We have no control over what the caller chooses to do. They could have a deferred `recover()`, even if it is ill-advised, and we should do our best to handle it.

Reply:
> We have no control over what the caller chooses to do

This is a fair concern and I updated the code to defer a call to `SQLiteStmt.finalize()` on exit since `finalize` is safe to multiple times and this will cover the case of the code panicking. 

As for overwriting the mutex we're safe here because nothing else is referencing the `SQLiteStmt` at this point in the code and if that were to change the race detector will catch it.